runtime/tchannel_client.go

```diff
@@ -43,150 +47,352 @@ type tchannelClient struct {
+		clientID:          opt.ClientID,
+		methodNames:       opt.MethodNames,
 		timeout:           opt.Timeout,
 		timeoutPerAttempt: opt.TimeoutPerAttempt,
+		loggers:           loggers,
+		metrics:           metrics,
 	}
-	return client
 }
 
-func (c *tchannelClient) writeArgs(call *tchannel.OutboundCall, headers map[string]string, req RWTStruct) error {
-	structWireValue, err := req.ToWire()
+// Call makes a RPC call to the given service.
+func (c *tchannelClient) Call(
+	ctx context.Context,
+	thriftService, methodName string,
+	reqHeaders map[string]string,
+	req, resp RWTStruct,
+) (success bool, resHeaders map[string]string, err error) {
+	retryOpts := tchannel.RetryOptions{
+		TimeoutPerAttempt: c.timeoutPerAttempt,
+	}
+	ctx, cancel := tchannel.NewContextBuilder(c.timeout).
+		SetParentContext(ctx).
+		SetRetryOptions(&retryOpts).
+		Build()
+	defer cancel()
+
+	serviceMethod := thriftService + "::" + methodName
+
+	call := &tchannelCall{
+		client:        c,
+		methodName:    c.methodNames[serviceMethod],
+		serviceMethod: serviceMethod,
+		reqHeaders:    reqHeaders,
+		logger:        c.loggers[serviceMethod],
+		metrics:       c.metrics[serviceMethod],
+	}
+	defer func() { call.finish(err) }()
+	call.start()
+
+	err = c.ch.RunWithRetry(ctx, func(ctx netContext.Context, rs *tchannel.RequestState) (cerr error) {
+		call.resHeaders, call.success = nil, false
+
+		call.call, cerr = c.sc.BeginCall(ctx, serviceMethod, &tchannel.CallOptions{
+			Format:       tchannel.Thrift,
+			RequestState: rs,
+		})
+		if cerr != nil {
+			call.logger.Error("Could not begin outbound request", zap.Error(cerr))
+			return errors.Wrapf(
+				err, "Could not begin outbound %s.%s (%s %s) request",
+				call.client.clientID, call.methodName, call.client.serviceName, call.serviceMethod,
+			)
+		}
+
+		// trace request
+		reqHeaders = tchannel.InjectOutboundSpan(call.call.Response(), reqHeaders)
+
+		if cerr := call.writeReqHeaders(reqHeaders); cerr != nil {
+			return cerr
+		}
+		if cerr := call.writeReqBody(req); cerr != nil {
+			return cerr
+		}
+
+		response := call.call.Response()
+		if cerr = call.readResHeaders(response); cerr != nil {
+			return cerr
+		}
+		if cerr = call.readResBody(response, resp); cerr != nil {
+			return cerr
+		}
+
+		return cerr
+	})
+
 	if err != nil {
-		return errors.Wrapf(err, "could not write request for outbound call %s: ", c.serviceName)
+		call.logger.Error("Could not make outbound request", zap.Error(err))
+		// Do not wrap system errors.
+		if _, ok := err.(tchannel.SystemError); ok {
+			return call.success, nil, err
+		}
+		return call.success, nil, errors.Wrapf(
+			err, "Could not make outbound %s.%s (%s %s) response",
+			call.client.clientID, call.methodName, call.client.serviceName, call.serviceMethod,
+		)
 	}
 
-	twriter, err := call.Arg2Writer()
+	return call.success, call.resHeaders, err
+}
+
+func (c *tchannelCall) start() {
+	c.startTime = time.Now()
+	c.metrics.Sent.Inc(1)
```

Comment:
we have not sent it yet here. Need to move this elsewhere.

Reply:
Will do.